### PR TITLE
Templatization of fixed-width parquet decoding kernels.

### PR DESCRIPTION
**Pull request by nvdbaranec**
This PR merges all of the fixed-width parquet decoding kernels into a single templatized kernel that can be selectively instantiated with desired features (dictionary/no-dictionary, nested/non-nested, etc).  It also adds support for (non-list) nested columns in this path. So structs do not have to use the much slower general decode kernel any more.

A new benchmark was added specific to structs containing only fixed width columns.  I added this because the performance improvement is fairly high (+20%) but we don't see it in the normal struct benchmarks because they include (and are dominated by) string decode times.  The new benchmark shows:

Before this PR:
```
| data_type |    io_type    | cardinality | run_length | bytes_per_second | peak_memory_usage | encoded_file_size |
|-----------|---------------|-------------|------------|------------------|-------------------|-------------------|
|    STRUCT | DEVICE_BUFFER |           0 |          1 |      21071216823 |         1.047 GiB |       511.675 MiB |
|    STRUCT | DEVICE_BUFFER |        1000 |          1 |      18974392387 |       821.312 MiB |       128.884 MiB |
|    STRUCT | DEVICE_BUFFER |           0 |         32 |      20429356824 |      621.787 MiB  |        28.141 MiB |
|    STRUCT | DEVICE_BUFFER |        1000 |         32 |      20572327813 |       598.421 MiB |        16.475 MiB |
```

After this PR:

```
| data_type |    io_type    | cardinality | run_length | bytes_per_second | peak_memory_usage | encoded_file_size |
|-----------|---------------|-------------|------------|------------------|-------------------|-------------------|
|    STRUCT | DEVICE_BUFFER |           0 |          1 |      25805996399 |         1.047 GiB |       511.675 MiB |
|    STRUCT | DEVICE_BUFFER |        1000 |          1 |      22422306660 |       821.312 MiB |       128.884 MiB |
|    STRUCT | DEVICE_BUFFER |           0 |         32 |      24460694014 |       621.787 MiB |        28.141 MiB |
|    STRUCT | DEVICE_BUFFER |        1000 |         32 |      24674861214 |       598.421 MiB |        16.475 MiB |
```

Split-page decoding for fixed-width types + structs are also going through this new path. New test added.

This brings us closer to eliminating the "general" kernel.  The only things left that run through it are lists and booleans.

This is PR 1 of 2, with the followup moving a lot of code around.  At this point, I think it makes sense to start consolidating our files a bit.

I also left some breadcrumbs (a few small commented out code blocks) in the core kernel `gpuDecodePageDataGeneric` for the next step of adding list support. They can be removed if people don't like them.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by nvdbaranec** [comment]
Leaving as a draft while spark tests cook.

File: cpp/src/io/parquet/decode_fixed.cu [hunk 1]
```diff
@@ -307,266 +186,293 @@ __device__ inline void gpuDecodeSplitValues(page_state_s* s,
...
+        thread_valid_count = thread_value_count;
+        block_valid_count  = block_value_count;
+      }
 
-    // only need to process definition levels if the column has nulls
-    if (nullable_with_nulls) {
-      processed_count += def_decoder.decode_next(t);
-      __syncthreads();
+      // if this is valid and we're at the leaf, output dst_pos
+      __syncthreads();  // handle modification of ni.value_count from below
```
Comment (hyperbolic2346) [inline cpp/src/io/parquet/decode_fixed.cu:301]: Is this and 308 copy/paste errors? `ni.valid_count` maybe? Why do we need both syncs for the same reason in the same code path?

**Review by hyperbolic2346 — COMMENTED** [review]
I like where this is going, but it's really hard to see what has actually changed due to the diffs. I have some questions, but I'm not entirely sure if they are new code.

File: cpp/src/io/parquet/decode_fixed.cu [hunk 2]
```diff
@@ -307,266 +186,293 @@ __device__ inline void gpuDecodeSplitValues(page_state_s* s,
...
+  int const last_row                  = first_row + s->num_rows;
+  int const capped_target_value_count = min(target_value_count, last_row);
 
-  if (!(BitAnd(pages[page_idx].kernel_mask, decode_kernel_mask::FIXED_WIDTH_NO_DICT))) { return; }
+  int const row_index_lower_bound = s->row_index_lower_bound;
 
-  // must come after the kernel mask check
-  [[maybe_unused]] null_count_back_copier _{s, t};
+  int const max_depth = s->col.max_nesting_depth - 1;
+  __syncthreads();
```
[the remark this answers is not in the document]
Comment (etseidl) [reply cpp/src/io/parquet/decode_fixed.cu:215]: I believe this is guarding `s->input_value_count` against a potential RAW (t0 writes it at L320).

File: cpp/src/io/parquet/decode_fixed.cu [hunk 3]
```diff
@@ -709,40 +720,113 @@ void __host__ DecodePageDataFixedDict(cudf::detail::hostdevice_span<PageInfo> pa
...
   dim3 dim_grid(pages.size(), 1);        // 1 thread block per page => # blocks
 
   if (level_type_size == 1) {
-    gpuDecodeSplitPageDataFlat<uint8_t><<<dim_grid, dim_block, 0, stream.value()>>>(
-      pages.device_ptr(), chunks, min_row, num_rows, error_code);
+    if (has_nesting) {
+      gpuDecodePageDataGeneric<uint8_t,
+                               decode_block_size,
+                               decode_kernel_mask::BYTE_STREAM_SPLIT_FIXED_WIDTH_NESTED,
+                               true,
```
Comment (etseidl) [inline cpp/src/io/parquet/decode_fixed.cu:795]: I think `has_dict_t` should be `false` for byte_stream_split, no?

File: cpp/src/io/parquet/parquet_gpu.hpp [hunk 4]
```diff
@@ -206,16 +206,20 @@ enum level_type {
...
+  GENERAL             = (1 << 0),  // Run catch-all decode kernel
+  STRING              = (1 << 1),  // Run decode kernel for string data
+  DELTA_BINARY        = (1 << 2),  // Run decode kernel for DELTA_BINARY_PACKED data
+  DELTA_BYTE_ARRAY    = (1 << 3),  // Run decode kernel for DELTA_BYTE_ARRAY encoded data
+  DELTA_LENGTH_BA     = (1 << 4),  // Run decode kernel for DELTA_LENGTH_BYTE_ARRAY encoded data
+  FIXED_WIDTH_NO_DICT = (1 << 5),  // Run decode kernel for fixed width non-dictionary pages
+  FIXED_WIDTH_DICT    = (1 << 6),  // Run decode kernel for fixed width dictionary pages
+  BYTE_STREAM_SPLIT   = (1 << 7),  // Run decode kernel for BYTE_STREAM_SPLIT encoded data
+  BYTE_STREAM_SPLIT_FIXED_WIDTH_FLAT = (1 << 8),  // Same as above but for flat, fixed-width data
+  BYTE_STREAM_SPLIT_FIXED_WIDTH_NESTED =
```
Comment (etseidl) [inline cpp/src/io/parquet/parquet_gpu.hpp:219]: Minor nit...`BYTE_STREAM_SPLIT` is only valid for fixed width, so adding `FIXED_WIDTH` is somewhat redundant. But I can see wanting to make it explicit.

File: cpp/src/io/parquet/decode_fixed.cu (hunk 1 above)
Comment (nvdbaranec) [reply cpp/src/io/parquet/decode_fixed.cu:301]: There's two cases:
- thread 1 could be reading ni.valid_count before the value stored on 311 has been written through from the previous loop.
- thread 1 could be reading ni.valid_count after thread 0 modifies it on line 311 on the current iteration of the loop

Comment (hyperbolic2346) [reply cpp/src/io/parquet/decode_fixed.cu:301]: This is more a challenge of the comments and not the syncthread calls.

Comment (nvdbaranec) [reply cpp/src/io/parquet/decode_fixed.cu:301]: Oh. Derp.

**Review by hyperbolic2346 — APPROVED** [review]

**Review by vuule — APPROVED** [review]
This one's really hard to follow using the diff.
Looks good, but that really surface-level with a diff like this.

File: cpp/src/io/parquet/reader_impl.cpp [hunk 5]
```diff
@@ -267,14 +267,26 @@ void reader::impl::decode_page_data(read_mode mode, size_t skip_rows, size_t num
...
+    DecodeSplitPageFixedWidthData(subpass.pages,
+                                  pass.chunks,
+                                  num_rows,
+                                  skip_rows,
+                                  level_type_size,
+                                  false,
+                                  error_code.data(),
+                                  streams[s_idx++]);
+  }
+
```
Comment (mhaseeb123) [inline cpp/src/io/parquet/reader_impl.cpp:280]: Minor: maybe we can add brief comments at each of these blocks like above.

File: cpp/src/io/parquet/decode_fixed.cu (hunk 2 above)
[the remark this answers is not in the document]
Comment (mhaseeb123) [reply cpp/src/io/parquet/decode_fixed.cu:215]: Minor: Wouldn't `__syncwarp()` be enough here as well?

File: cpp/src/io/parquet/decode_fixed.cu [hunk 6]
```diff
@@ -307,266 +186,293 @@ __device__ inline void gpuDecodeSplitValues(page_state_s* s,
...
-                     s->abs_lvl_start[level_type::DEFINITION],
-                     s->abs_lvl_end[level_type::DEFINITION],
-                     def,
-                     s->page.num_input_values);
-  }
-  __syncthreads();
+      // thread and block validity count
+      int thread_valid_count, block_valid_count;
+      if constexpr (nullable) {
+        using block_scan = cub::BlockScan<int, decode_block_size>;
```
Comment (mhaseeb123) [inline cpp/src/io/parquet/decode_fixed.cu:255]: Curious: Would the `cub::block_scan.Inclusivesum` with cooperative groups be simpler and more readable?

```cpp
    // I think it should not cause any race conditions as block_valid_count is passed as rvalue as per inclusive_scan API signature.
    block_valid_count = cg::inclusive_scan(cg::this_thread_block(), block_valid_count, cg::plus<int>()); 
```

Comment (nvdbaranec) [reply cpp/src/io/parquet/decode_fixed.cu:255]: Potentially. I am not particularly familiar with CG.  Wouldn't this require some additional setup in the kernel itself before it could be used?

Comment (mhaseeb123) [reply cpp/src/io/parquet/decode_fixed.cu:255]: Afaik, just the following two lines at the top of the file and nothing else in the kernel. Maybe @PointKernel can better comment. The CUB version is perfectly fine though.

```cpp
#include <cooperative_groups>
namespace cg = cooperative_groups;
```

Comment (PointKernel) [reply cpp/src/io/parquet/decode_fixed.cu:255]: Per https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#inclusive-scan-and-exclusive-scan, `inclusive_scan` only works with `coalesced_group` and `thread_block_tile` but not `thread_block`. CUB scan is probably the best option for now. Once it's available, `inclusive_scan` is definitely a much cleaner solution.

**Review by mhaseeb123 — APPROVED** [review]
Thanks for the effort. Couldn't review too deep due to the complex diff as others noted.

File: cpp/src/io/parquet/decode_fixed.cu (hunk 6 above)
Comment (mhaseeb123) [reply cpp/src/io/parquet/decode_fixed.cu:255]: Thanks for clarifying @PointKernel !

File: cpp/src/io/parquet/decode_fixed.cu (hunk 3 above)
Comment (etseidl) [inline cpp/src/io/parquet/decode_fixed.cu:795]:
```suggestion
                               false,
```

File: cpp/src/io/parquet/decode_fixed.cu [hunk 7]
```diff
@@ -709,40 +720,113 @@ void __host__ DecodePageDataFixedDict(cudf::detail::hostdevice_span<PageInfo> pa
...
+                               true,
+                               true,
+                               decode_fixed_width_split_values_func>
+        <<<dim_grid, dim_block, 0, stream.value()>>>(
+          pages.device_ptr(), chunks, min_row, num_rows, error_code);
+    } else {
+      gpuDecodePageDataGeneric<uint8_t,
+                               decode_block_size,
+                               decode_kernel_mask::BYTE_STREAM_SPLIT_FIXED_WIDTH_FLAT,
+                               true,
```
Comment (etseidl) [inline cpp/src/io/parquet/decode_fixed.cu:804]:
```suggestion
                               false,
```

File: cpp/src/io/parquet/decode_fixed.cu [hunk 8]
```diff
@@ -709,40 +720,113 @@ void __host__ DecodePageDataFixedDict(cudf::detail::hostdevice_span<PageInfo> pa
...
+          pages.device_ptr(), chunks, min_row, num_rows, error_code);
+    }
   } else {
-    gpuDecodeSplitPageDataFlat<uint16_t><<<dim_grid, dim_block, 0, stream.value()>>>(
-      pages.device_ptr(), chunks, min_row, num_rows, error_code);
+    if (has_nesting) {
+      gpuDecodePageDataGeneric<uint16_t,
+                               decode_block_size,
+                               decode_kernel_mask::BYTE_STREAM_SPLIT_FIXED_WIDTH_NESTED,
+                               true,
```
Comment (etseidl) [inline cpp/src/io/parquet/decode_fixed.cu:815]:
```suggestion
                               false,
```

File: cpp/src/io/parquet/decode_fixed.cu [hunk 9]
```diff
@@ -709,40 +720,113 @@ void __host__ DecodePageDataFixedDict(cudf::detail::hostdevice_span<PageInfo> pa
...
+                               true,
+                               decode_fixed_width_split_values_func>
+        <<<dim_grid, dim_block, 0, stream.value()>>>(
+          pages.device_ptr(), chunks, min_row, num_rows, error_code);
+    } else {
+      gpuDecodePageDataGeneric<uint16_t,
+                               decode_block_size,
+                               decode_kernel_mask::BYTE_STREAM_SPLIT_FIXED_WIDTH_FLAT,
+                               true,
```
Comment (etseidl) [inline cpp/src/io/parquet/decode_fixed.cu:824]:
```suggestion
                               false,
```

**Review by etseidl — CHANGES_REQUESTED** [review]
Some suggestions that were marked resolved but maybe got lost in the shuffle. Should save a bunch of shared mem for the byte stream split decoder.

File: cpp/src/io/parquet/decode_fixed.cu (hunk 3 above)
Comment (nvdbaranec) [reply cpp/src/io/parquet/decode_fixed.cu:795]: Shoot. I missed these somehow.  I'll put up a followup.  I could swear I pushed these changes.
